src/RequestExecutor/HttpRequestExecutor.ts

```diff
@@ -474,4 +468,84 @@ export class HttpRequestExecutor implements RequestExecutor {
 
     return new Request(result);
   }
+
+  private certificatesForRequest(request: Request): Cert[] {
+    const cachedCertificate = this.certificatesCache.get(request);
+    if (cachedCertificate) {
+      return [cachedCertificate];
+    }
+
+    const requestUrl = new URL(request.url);
+    const port = Helpers.portFromURL(requestUrl);
+
+    return this.options.certs.filter((cert: Cert) =>
+      Helpers.matchHostnameAndPort(requestUrl.hostname, port, cert)
+    );
+  }
```

Comment:
This is duplicated in multiple places across the codebase. Consider to extract and reuse